storage/limit.go

```diff
@@ -0,0 +1,36 @@
+package storage
+
+import (
+	"errors"
+
+	"github.com/go-git/go-git/v5/plumbing"
+)
+
+var ErrLimitExceeded = errors.New("limit exceeded")
+
+// Limited wraps git.Storer to limit the number of bytes that can be stored.
+type Limited struct {
+	Storer
+	N *int64
+}
+
+// Limit returns a git.Storer limited to the specified number of bytes.
+func Limit(s Storer, n int64) *Limited {
+	return &Limited{Storer: s, N: &n}
+}
+
+func (s *Limited) SetEncodedObject(obj plumbing.EncodedObject) (plumbing.Hash, error) {
+	if *s.N - obj.Size() < 0 {
+		return plumbing.ZeroHash, ErrLimitExceeded
+	}
+	*s.N -= obj.Size()
```

Comment:
[Atomic](https://pkg.go.dev/sync/atomic) counters would remove concurrency concerns here.

I understand that there are [concurrency issues](https://github.com/go-git/go-git/issues/773) within go-git already, but it would be better if new code was not affected by them as well.